src/components/advanced-color-picker/advanced-color-picker.test.tsx

```diff
@@ -0,0 +1,309 @@
+  { value: "#A3CAF0", label: "blue" },
+  { value: "#FD9BA3", label: "pink" },
+  { value: "#B4AEEA", label: "purple" },
+  { value: "#ECE6AF", label: "goldenrod" },
+  { value: "#EBAEDE", label: "orchid" },
+  { value: "#EBC7AE", label: "desert" },
+  { value: "#AEECEB", label: "turquoise" },
+  { value: "#AEECD6", label: "mint" },
+];
+
+const ControlledColorPicker = (props: Partial<AdvancedColorPickerProps>) => {
+  const [color, setColor] = useState<string | undefined>();
+
+  return (
+    <AdvancedColorPicker
+      name="advancedPicker"
+      defaultColor=""
+      availableColors={availableColors}
+      selectedColor={color}
+      onChange={(event) => setColor(event.target.value)}
+      {...props}
+    />
+  );
+};
+
+beforeAll(() => {
+  jest.useFakeTimers();
+});
+
+afterAll(() => {
+  jest.useRealTimers();
+});
+
+testStyledSystemMargin((props) => (
+  <ControlledColorPicker name="advancedPicker" {...props} />
+));
+
+test("should display deprecation warning once when rendered as uncontrolled", () => {
+  const loggerSpy = jest
+    .spyOn(Logger, "deprecate")
+    .mockImplementation(() => {});
+  render(
+    <AdvancedColorPicker
+      name="advancedPicker"
+      availableColors={availableColors}
+      defaultColor="#EBAEDE"
+    />
+  );
+
+  expect(loggerSpy).toHaveBeenCalledWith(
+    "Uncontrolled behaviour in `Advanced Color Picker` is deprecated and support will soon be removed. Please make sure all your inputs are controlled."
+  );
+  expect(loggerSpy).toHaveBeenCalledTimes(1);
+
+  loggerSpy.mockRestore();
+});
+
+test("when the main button is clicked, the dialog opens and the onOpen callback is called", async () => {
+  const user = userEvent.setup({ advanceTimers: jest.advanceTimersByTime });
+
+  const onOpen = jest.fn();
+
+  render(<ControlledColorPicker onOpen={onOpen} />);
+  expect(screen.queryByRole("dialog")).not.toBeInTheDocument();
+  expect(onOpen).not.toHaveBeenCalled();
+
+  await user.click(screen.getByRole("button", { name: "Change colour" }));
+
+  expect(screen.getByRole("dialog")).toBeVisible();
+  expect(onOpen).toHaveBeenCalledTimes(1);
+});
+
+test.each([
+  ["enter", "{Enter}"],
+  ["space", " "],
+])("when the %s key is pressed, the dialog should open", async (_, keyCode) => {
+  const user = userEvent.setup({ advanceTimers: jest.advanceTimersByTime });
+  render(<ControlledColorPicker />);
+  screen.getByRole("button", { name: "Change colour" }).focus();
+  await user.keyboard(keyCode);
+  expect(screen.getByRole("dialog")).toBeVisible();
+});
+
+test.each(["a", "b", "q", "t", "x", "4", "0"])(
+  "when the %s key is pressed, the dialog should not open",
+  async (key) => {
+    const user = userEvent.setup({ advanceTimers: jest.advanceTimersByTime });
+    render(<ControlledColorPicker />);
+    screen.getByRole("button", { name: "Change colour" }).focus();
+    await user.keyboard(key);
+    expect(screen.queryByRole("dialog")).not.toBeInTheDocument();
```

Comment:
I agree with this suggestion for tests where the dialog is closed through some user interaction, and will add it to them 👍 

However, I'm not going to make this change to this specific place you've highlighted it here - this is because that particular test is simply testing that the dialog, which is currently closed, does not open in response to specific key presses. In this case there is no need to wait for the dialog not being in the document because it already isn't, and in my opinion the test would look rather odd (waiting for something that is, by design, already the case even before the action).